reduce_confusion_matrix.py

```diff
@@ -0,0 +1,116 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# ***************************************************************************80
+#
+# reduce_confusion_matrix.py -f <input_filename> -c <no of top mispredictions>
+#
+# *****************************************************************************
+
+# standard imports
+import heapq
+
+# third party imports
+import pandas
+import click
+from sklearn.metrics import confusion_matrix
+
+# Typical phrases.csv from HF contains
+# Labelled Phrase,Detected Phrase,Intent Id,Intent Name,Top Match Intent Id,Top Match Intent Name,Top Match Score,Entropy,Uncertainty,Margin Score,Result Type
+
+@click.command()
+@click.option('-m', '--top_mispredictions', type=int, default=5, help='Number of top mispredictions')
+@click.option('-f', '--phrases_filename', type=str, default='./data/phrases.csv', help='Phrases')
+def main(phrases_filename: str, top_mispredictions: int) -> None:
+    '''Main function'''
+
+    reduce_confusion_matrix(phrases_filename, top_mispredictions)
+
+def reduce_confusion_matrix(phrases_filename: str, top_mispredictions: int) -> None:
+    '''Reduces Confusion matrix'''
+    df = pandas.read_csv(phrases_filename)
+    
+    # summarise
+    print('Summary of df')
+    print(df[["Intent Name","Result Type"]].groupby(["Result Type"]).count())
+
+    # create confusion matrix
+    labels = sorted(list(set(df["Intent Name"])))
+    print(f"Total Number of intents: {len(labels)}")
+    matrix = confusion_matrix(df["Intent Name"],df["Top Match Intent Name"],labels = labels)
+    
+    # replace true positives to 0 and sum all the incorrect predictions
+    total_mispredictions = 0
+    for i in range(len(matrix)):
+        for j in range(len(matrix[i])):
+            if i == j:
+                matrix[i][j] = 0
+            else:
+                total_mispredictions = total_mispredictions + matrix[i][j]
+    print("\nTrue positives are replaced with 0\n")
+    
+    # confusion matrix into dataframe
+    df_matrix = pandas.DataFrame(data=matrix,index=labels,columns=labels)
+
+    # get the list of top mispredictions e.g. top 10 largest number of incorrect predictions
+    top_mispredictions_list = (heapq.nlargest(top_mispredictions,matrix.max(1)))
```

Comment:
heapq new on me, if I understand we aren't using the heap object, just the nlargest.  I can't see an obvious alternative as clean as this.   But would like to see a test to verify that it's correct and the special heap properties don't cause issues.